### ISSUE TITLE
All DSP objects should use `class_setdspflags` for new DSP flags

### ISSUE THREAD

**Issue by Spacechild1** [pure-data/pure-data#2035]
Currently, some DSP objects pass the new DSP flags (`CLASS_MULTICHANNEL`, `CLASS_NOPROMOTELEFT` and `CLASS_NOPROMOTESIGN`) to the new function `class_setdspflags`, but most others just pass them to `class_new`.

I think we should really encourage the former - and maybe even *disallow* the latter!

Here's why: if an (external) object does not call `class_setdspflags`, it would load successfully in older Pd versions and crash horribly. Enforcing the use of `class_setdspflags` makes sure that this can never happen. (Multichannel DSP objects *with signal outputs* are safe because they must call the new `signal_setmultiout` function, but this is just a particular subset.)

Any thoughts?

**Comment by umlaeute** [comment]
well, if you use `class_setdspflags`, than the external will only load in a Pd>=0.54 (since this is when the symbol appeared).

using `class_new()` should make it possible to handle older Pd versions within the code (where this is feasible, obviously. e.g. an object that like the binops that doesn't do any fancy  "cross-talk" between the multichannels).

or am i missing something?

i agree that devs should be aware of the implications.

**Comment by Spacechild1** [comment]
> well, if you use class_setdspflags, than the external will only load in Pd>=0.54 (since this is when the symbol appeared).

That's the point :)

> using class_new() should make it possible to handle older Pd versions within the code

Well, if you set the `CLASS_MULTICHANNEL` flag, you are saying: this object is multichannel aware and consequently it will *only* work correctly in Pd >= 0.54.

If the object has signal outputs, the objects must call `signal_setmultiout` anyway. Either way, it must access the new `s_nchans` member of `t_signal`, which only works in Pd >= 0.54.

>  (where this is feasible, obviously. e.g. an object that like the binops that doesn't do any fancy "cross-talk" between the multichannels).

This doesn't work, see above.

**Comment by umlaeute** [comment]
> If you set the `CLASS_MULTICHANNEL` flag, the object will need to access the new `s_nchans` member in `t_signal` and thus only work correctly in Pd >= 0.54.

what i meant is:
1. you declare `CLASS_MULTICHANNEL` in `pd_class()`
2. in the `dsp` callback, you do a runtime check:
    ```C
    unsigned int nchans = (pd_version >= 0.54)?sp[0]->s_n_chans:1;
    ```

> This doesn't work. `s_length` is the length of a channel, not the total buffer length!

well, in Pd<0.54, `s_nchans` is implicitly **1**, and thus `s_length` is both the  length of a channel *and* the total buffer length!

i  haven't thought about the use cases of `CLASS_NOPROMOTELEFT` and `CLASS_NOPROMOTESIG` for Pd<0.54

**Comment by Spacechild1** [comment]
> well, in Pd<0.54, s_nchans is implicitly 1

In  Pd<0.54 `s_nchans` does not exist, you would accidentally read `s_refcount` instead (which should be 0, see `ugen_stop`).

s_refcount

**Comment by Spacechild1** [comment]
Ah sorry, I missed the runtime check in your example.

**Comment by Spacechild1** [comment]
Ok, I think the only problematic case is multichannel DSP objects *without* signal outlets, but these can be handled with a runtime check. I could add a comment to `m_pd.h` (https://github.com/pure-data/pure-data/blob/f2b7874669bb7c7cbe03afeb6556068c1c2ce6c7/src/m_pd.h#L477-L484).

**Comment by Spacechild1** [comment]
Since multichannel DSP objects *with* signal outlets don't load in older Pd versions anyway, I might just write something like this to keep it simple:
```
Classes registered with CLASS_NOPROMOTESIG and CLASS_NOPROMOTELEFT also work in Pd < 0.54, so you may just pass these flags to class_new().

Classes registered with CLASS_MULTICHANNEL only work in Pd >= 0.54. In this case it is recommended to set *all* DSP flags with class_setdspflags() - instead of passing them to class_new() - to prevent the external from loading in older Pd versions.
```

("experts" will know that you can also check the Pd version at runtime. You may even use `dlsym` resp. `GetProcAddress` to get the `signal_setmultiout` function at runtime and adapt your code accordingly; that's what I'm planning to do with `vstplugin~` to keep it running in Pd < 0.54.)

**Comment by millerpuckette** [comment]
I think the flages are already usable from class_new() so there's no real need for class_setdspflags().  Perhaps this means I should remove the function and change everything over to class_new().  I thought it made the code more readable to call a separate function but now I can't see why that would be true :)

**Comment by Spacechild1** [comment]
Yeah, I think `class_setdspflags` seems redundant and could be removed altogether.

`class_getdspflags`, on the other hand, can be very useful, but IMO it should be rather called `class_getflags` and return *all* flags. (We might add more flags in the future and not all of them would necessarily be "DSP flags".) An external can use this function to easily check if all the flags it has passed to `class_new` are actually supported and provide necessary fallbacks.

It can also be used to prevent multichannel DSP objects *without* signal outputs from loading in older Pd versions:
```
void foo_tilde_setup(void)
{
        /* "foo" has no signal outputs. Since it doesn't call signal_setmultiout,
        it would load in older Pd versions, but crash at runtime when accessing
        and using `s_nchans` in `t_signal`. */
    foo_class = class_new("foo", (t_newmethod)foo_new,
        (t_method)foo_free, sizeof(t_foo), CLASS_MULTICHANNEL, 0);

        /* actually, this external wouldn't even load in Pd < 0.54 */
    if ((class_getflags(foo_class) & CLASS_MULTICHANNEL) == 0)
    {
        pd_error(0, "no multichannel support");
        return;
    }
}
```

**Comment by millerpuckette** [comment]
Aha.. that's why I had added the class_setdspflags() function and started using it... so that extern writers might use it too and prevent their externs from loading in older Pd versions.  So... perhaps it's OK to just leave it in :)
OTOH I don't think class_getdspflags() is a useful idea... externs can just check the Pd version.

**Comment by Spacechild1** [comment]
> so that extern writers might use it too and prevent their externs from loading in older Pd versions.

Yes, that's what I figured. However, that would only work if they *had* to use these functions.

Since we have to *tell* external authors what to do, we might just as well tell them to do a runtime check. At least it would be more obvious what's going on.

IMO `class_setdspflags` is just confusing ("Why do most objects pass their flags to class_new(), but some use class_setdspflags()!?") Also, I don't think it is good to have a function that is completely redundant and its *only* purpose is to prevent an external from loading below a particular Pd version.

Note that the problem *only* exists with multichannel DSP objects *without* signal outputs. Multichannel DSP objects *with* signal outputs wouldn't load anyway because they have to call `signal_setmultiout`. Objects that set `CLASS_NOPROMOTESIG` and/or `CLASS_NOPROMOTELEFT` still work in Pd < 0.54.

---

Side note: can we maybe add something like `sys_checkversion(int major, int minor, int bugfix, int greaterequal)` to simplify runtime version checks?

**Comment by Spacechild1** [comment]
If you are more into hacks, we could also allow to pass to `NULL` to `signal_setmultiout`:

```
void foo_dsp(t_foo *x, t_signal **sp)
{
        /* t_foo is multichannel, but has no signal outputs; this prevents it from loading in Pd < 0.54 */
    signal_setmultiout(NULL, 0);
        /* ... */
}
```

I would find this less awkward than `class_setdspflags`.

**Comment by millerpuckette** [comment]
yep, a runtime check for version >= 0.54 in the setup routine would do it.  OK, I'm taking the offending function out :)
